src/main/java/com/android/volley/toolbox/DiskBasedAsyncCacheFallback.java

```diff
@@ -0,0 +1,454 @@
+import java.io.BufferedInputStream;
+import java.io.DataInputStream;
+import java.io.EOFException;
+import java.io.File;
+import java.io.FileInputStream;
+import java.io.FileNotFoundException;
+import java.io.FilterInputStream;
+import java.io.IOException;
+import java.io.InputStream;
+import java.io.OutputStream;
+import java.util.ArrayList;
+import java.util.Collections;
+import java.util.LinkedHashMap;
+import java.util.List;
+import java.util.Map;
+
+public class DiskBasedAsyncCacheFallback extends AsyncCache {
+
+    /** Map of the Key, CacheHeader pairs */
+    private final Map<String, DiskBasedAsyncCacheFallback.CacheHeader> mEntries =
+            new LinkedHashMap<>(16, .75f, true);
+
+    /** Total amount of space currently used by the cache in bytes. */
+    private long mTotalSize = 0;
+
+    /** The supplier for the root directory to use for the cache. */
+    private final DiskBasedAsyncCacheFallback.FileSupplier mRootDirectorySupplier;
+
+    /** The maximum size of the cache in bytes. */
+    private final int mMaxCacheSizeInBytes;
+
+    /** Default maximum disk usage in bytes. */
+    private static final int DEFAULT_DISK_USAGE_BYTES = 5 * 1024 * 1024;
+
+    /** High water mark percentage for the cache */
+    @VisibleForTesting static final float HYSTERESIS_FACTOR = 0.9f;
+
+    /** Magic number for current version of cache file format. */
+    private static final int CACHE_MAGIC = 0x20150306;
+
+    /**
+     * Constructs an instance of the DiskBasedAsyncCacheFallback at the specified directory.
+     *
+     * @param rootDirectory The root directory of the cache.
+     * @param maxCacheSizeInBytes The maximum size of the cache in bytes. Note that the cache may
+     *     briefly exceed this size on disk when writing a new entry that pushes it over the limit
+     *     until the ensuing pruning completes.
+     */
+    public DiskBasedAsyncCacheFallback(final File rootDirectory, int maxCacheSizeInBytes) {
+        mRootDirectorySupplier =
+                new DiskBasedAsyncCacheFallback.FileSupplier() {
+                    @Override
+                    public File get() {
+                        return rootDirectory;
+                    }
+                };
+        mMaxCacheSizeInBytes = maxCacheSizeInBytes;
+    }
+
+    /** Clears the cache. Deletes all cached files from disk. */
+    @Override
+    public synchronized void clear() {
+        // TODO (sphill99): Implement
+    }
+
+    /** Returns the cache entry with the specified key if it exists, null otherwise. */
+    @Override
+    public synchronized void get(String key, OnGetCompleteCallback callback) {
+        DiskBasedAsyncCacheFallback.CacheHeader entry = mEntries.get(key);
+        // if the entry does not exist, return.
+        if (entry == null) {
+            callback.onGetComplete(null);
+        }
+        File file = getFileForKey(key);
+
+        try {
+            DiskBasedAsyncCacheFallback.CountingInputStream cis =
```

Comment:
I believe you can use a try-with-resources here as well.

Reply:
It says that it requires an API of 19 or higher to do a try-with-resources.